File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        res = 0
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        visited = set()
        
        for i, row in enumerate(grid):
            for j, col in enumerate(row):
                if grid[i][j] == "1" and (i,j) not in visited:
                    res += 1
                    stack = deque([(i,j)])
                    visited.add((i,j))
                    
                    while stack:
                        curr_i, curr_j = stack.popleft()
                        
                        for d in directions:
                            new_i = curr_i + d[0]
                            new_j = curr_j + d[1]
                            
                            if 0 <= new_i < len(grid) and 0 <= new_j < len(grid[0]):
                                if grid[new_i][new_j] == "1" and (new_i, new_j) not in visited:
                                    
                                    stack.append((new_i, new_j))
                                    visited.add((new_i, new_j))
        return res
                            
                        
                    

File: 0200-number-of-islands/0200-number-of-islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        
        int res = 0;
        HashSet<(int, int)> seen = new HashSet<(int, int)>();
        int[][] directions = new int[4][] {
            new int[] {0, 1},
            new int[] {0, -1},
            new int[] {1, 0},
            new int[] {-1, 0}
        };

        for (int i = 0; i < grid.Length; i++){
            for (int j = 0; j < grid[i].Length; j++){
                if(grid[i][j] == '1' && !seen.Contains((i,j))){
                    res++;

                    Queue<(int, int)> queue = new Queue<(int, int)>();
                    queue.Enqueue((i,j));

                    while(queue.Count > 0){
                        var curr = queue.Dequeue();
                        if(seen.Contains(curr)) continue;
                        seen.Add(curr);
                        foreach (var dir in directions){
                            int newX = curr.Item1 + dir[0];
                            int newY = curr.Item2 + dir[1];
                            if (newX >= 0 && newX < grid.Length && 
                            newY >= 0 && newY < grid[newX].Length && 
                            grid[newX][newY] == '1' && !seen.Contains((newX,newY))){
                                queue.Enqueue((newX, newY));
                            }
                        }
                    }
                }
            }
        }
        return res;
    }
}